### ISSUE TITLE
When adding a custom metric to AutoSklearnClassifier the run crashes

### ISSUE THREAD

**Issue by chogovadze** [automl/auto-sklearn#1449]
I would like to start by thanking the team for the very interesting project!

When creating a custom metric in the following fashion: 

```python
def myaccu(solution, prediction):
    return np.mean(solution == prediction)

accu = autosklearn.metrics.make_scorer(
    name="myacc",
    score_func=myaccu,
    optimum=1,
    greater_is_better=True,
    needs_proba=False,
    needs_threshold=False,
)

model = autosklearn.classification.AutoSklearnClassifier(time_left_for_this_task=..., 
                                                       per_run_time_limit=5*60, 
                                                       resampling_strategy=120,
                                                       n_jobs=8,
                                                       ensemble_size=10,
                                                       metric=accu)

```

the run crashes with the output:

`Dummy prediction failed with run state StatusType.CRASHED and additional output: {'error': 'Result queue is empty', 'exit_status': "<class 'pynisher.limit_function_call.AnythingException'>", 'subprocess_stdout': None, 'subprocess_stderr': None, 'exitcode': 1, 'configuration_origin': 'DUMMY'}`

If the function `myaccu` is replaced with `autosklearn.metrics.accuracy` then there are no issues. 

However, the following configuration does not work:

```python
def myaccu(solution, prediction):
    return autosklearn.metrics.accuracy(solution, prediction)
```

Considering that I am simply wrapping the function I would expect this to still work. 

Thank you for your time!

- Linux
- Python 3.7.4
- Auto-sklearn 0.14.6

**Comment by Eslam-Zaher** [comment]
Have you worked it out? Any update on how to solve this issue?

**Comment by chogovadze** [comment]
@Eslam-Zaher unfortunately, not yet

**Comment by eddiebergman** [comment]
Hi @Eslam-Zaher and @chogovadze,

The issue seems to be coming with how new subprocesses are spawned, specifically the context in which they spawn. There was an important detail in the longer error message I got but I'll add some notes at the bottom here as to what causes this.
```
Can't pickle <function myaccu at 0x7fd79f688310>: it's not the same object as __main__.myaccu`
```

### Solution
I did some testing and the solution is not the nicest but worked, essentially move the metric to another file. 
```python
# other.py
import numpy as np

def customacc(solution, prediction):
    return np.mean(solution == prediction)
```

```python
# runmy.py
from autosklearn.pipeline.util import get_dataset
from autosklearn.metrics import make_scorer, accuracy
from autosklearn.classification import AutoSklearnClassifier


from other import customacc

if __name__ == "__main__":

    accu = make_scorer(
        name="myacc",
        score_func=customacc,
        optimum=1,
        greater_is_better=True,
        needs_proba=False,
        needs_threshold=False,
    )

    model = AutoSklearnClassifier(
        time_left_for_this_task=60,
        per_run_time_limit=5,
        resampling_strategy='holdout',
        n_jobs=2,
        ensemble_size=10,
        metric=accu,
    )

    X, y, Xt, yt = get_dataset("iris")

    model.fit(X, y)

    print(model.score(Xt, yt))
```

---

The exact reason, I'm not sure but there 2 main points for your own information and for our own tracking:
1. When you use `njob=1`, the error didn't occur so there's a line in the code that triggered some suspicion 
https://github.com/automl/auto-sklearn/blob/5078e34537d9f354f89df84f6557922cc1283a09/autosklearn/automl.py#L262-L265
These suspicions were confirmed as when I changed the source code to just use `fork` all the time, there were no problems. I won't give an overview of the difference but here is some documentation:
* [python.multiprocessing](https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods)
* [A nice blog](http://www.bnikolic.co.uk/blog/python/parallelism/2019/11/13/python-forkserver-preload.html) post on explaining the differences.
The blog post is short but gives "context" (heh heh) to the reason, _it seems "forkserver" doesn't pickle the `__main__.myaccu` in a way that it can be reconstructed in the subprocess. I also remember seeing some error message where the subprocess was working in the namespace `__mp_main__` which might add to the issue.
The reason this doesn't happen with data like `X` and `y` is because we save those to disk and load them up in the subprocess, i.e. there's no pickled function definition.
2. Pynisher consuming error messages is very annoying to debug why things went wrong and in this case would at least give more context as to what was going wrong. This [PR for `v0.7.0`](https://github.com/automl/pynisher/pull/18) should hopefully relieve the issue.

---

I will leave this issue open for 2-3 days and discuss this with @mfeurer as to why we use `forkserver` vs `fork`. I imagine it's safer at shutting down subproccess incase of crashes. There's not much we can really do here in the short term so I would close the issue after that if it solves your issue.

Best,
Eddie

**Comment by mfeurer** [comment]
We discuss the need for forkserver/spwan in [the pynisher docs](https://github.com/automl/pynisher#pynisher-and-multithreading). We decided to use the forkserver over spawn as it reduces the subprocess startup time of <0.1s compared to >1 for spawn.

**Comment by chogovadze** [comment]
@eddiebergman Thank you for the explanation!
